### ISSUE TITLE
Dockerized Pihole not starting if ServerIP is non-local (HA/keepalived use-case)

### ISSUE THREAD

**Issue by lukaszczerpak** [pi-hole/docker-pi-hole#855]
### Versions
Pi-hole version is v5.3.1 (Latest: v5.3.1)
AdminLTE version is v5.5 (Latest: v5.5)
FTL version is v5.8.1 (Latest: v5.8.1)

### Platform
- OS and version:  Arch Linux
- Platform:  Docker

### Expected behavior

Pihole should be able to bind to non-local IP address (that was working in versions <=5.7).

```
Setting DNS servers based on PIHOLE_DNS_ variable
Custom WEB_PORT set to 3180
INFO: Without proper router DNAT forwarding to 192.168.5.3:3180, you may not get any blocked websites on ads
::: Pre existing WEBPASSWORD found
DNSMasq binding to custom interface: dmz
Added ENV to php:
			"PHP_ERROR_LOG" => "/var/log/lighttpd/error.log",
			"ServerIP" => "192.168.5.3",
			"VIRTUAL_HOST" => "192.168.5.3",
Using IPv4
::: Preexisting ad list /etc/pihole/adlists.list detected ((exiting setup_blocklists early))
https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
https://mirror1.malwaredomains.com/files/justdomains
https://s3.amazonaws.com/lists.disconnect.me/simple_tracking.txt
https://s3.amazonaws.com/lists.disconnect.me/simple_ad.txt
::: Testing pihole-FTL DNS: FTL started!
::: Testing lighttpd config: Syntax OK
::: All config checks passed, cleared for startup ...
::: Enabling Query Logging
  [i] Enabling logging...
  [✓] Logging has been enabled!
 ::: Docker start setup complete
  Skipping Gravity Database Update.
  Pi-hole version is v5.3.1 (Latest: v5.3.1)
  AdminLTE version is v5.5 (Latest: v5.5)
  FTL version is v5.8.1 (Latest: v5.8.1)
[cont-init.d] 20-start.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
Starting crond
Starting lighttpd
Starting pihole-FTL (no-daemon) as root
[services.d] done.
```

### Actual behavior / bug

Pihole fails to start and throws the following errors:

```
Setting DNS servers based on PIHOLE_DNS_ variable
Custom WEB_PORT set to 3180
INFO: Without proper router DNAT forwarding to 192.168.5.3:3180, you may not get any blocked websites on ads
::: Pre existing WEBPASSWORD found
DNSMasq binding to custom interface: dmz
Added ENV to php:
			"PHP_ERROR_LOG" => "/var/log/lighttpd/error.log",
			"ServerIP" => "192.168.5.3",
			"VIRTUAL_HOST" => "192.168.5.3",
Using IPv4
::: Preexisting ad list /etc/pihole/adlists.list detected ((exiting setup_blocklists early))
https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
https://mirror1.malwaredomains.com/files/justdomains
https://s3.amazonaws.com/lists.disconnect.me/simple_tracking.txt
https://s3.amazonaws.com/lists.disconnect.me/simple_ad.txt
sh: 1: addr2line: not found
sh: 1: addr2line: not found
sh: 1: addr2line: not found
sh: 1: addr2line: not found
sh: 1: addr2line: not found
sh: 1: addr2line: not found
```

### Steps to reproduce

1. Setup Linux to enable bind to non-local IP:
```
$ cat /etc/sysctl.d/99-vrrp.conf 
net.ipv4.ip_nonlocal_bind=1
net.ipv6.ip_nonlocal_bind=1
```
2. Start docker container (ver. >=5.8) with host mode and set ServerIP to address that matches network on an interface (ie. 192.168.5.0/24) but is not configured (192.168.5.3)
3. Observe errors in the container output
4. Stop and remove the container
5. Start docker container based on older image (ver. <=5.7) with host mode and set ServerIP to address that matches network on an interface (ie. 192.168.5.0/24) but is not configured (192.168.5.3)
6. Observe all works fine

### Additional context

Please let me know if any additional information or tests would be needed.

**Comment by lukaszczerpak** [comment]
the output is from `docker logs pihole` - it's related to one of commands used during startup. Tried to track it down but frankly got stuck and I am out of ideas. Looks like some underlying exec in docker container, so I don't think it's Arch.

**Comment by lukaszczerpak** [comment]
However my backup machine is Raspbian - I can do the same exercise to see if the behavior is the same. 2mins...

**Comment by lukaszczerpak** [comment]
> 

I tried with no config files, except this one I need due to custom setup I have - `dnsmasq.d/99-custom-dns.conf`:

```
listen-address=192.168.5.3
except-interface=*
bind-interfaces
```

The same errors were thrown.

**Comment by dschaper** [comment]
Try without any files at all. The goal is to see if a plain, unconfigured container starts.

**Comment by lukaszczerpak** [comment]
> I don't know why the underlying OS would have any affect but maybe https://aur.archlinux.org/packages/pi-hole-server/?O=10&PP=10 has something?

Dockerized version of Pihole should be host OS agnostic imho. What is bugging me is that the same setup but using v5.7 works like a charm. I will test it more as I am pretty surprised and excited it works well on Raspbian...

**Comment by dschaper** [comment]
> Dockerized version of Pihole should be host OS agnostic imho.

I completely agree. But that is more dependent upon the underlying OS interacting with Docker.



> I will test it more as I am pretty surprised and excited it works well on Raspbian...

I'm not sure what that should be surprising?

**Comment by lukaszczerpak** [comment]
I installed `binutils` inside the container to provide with `addr2line` command and now I am getting different errors:

```
addr2line: 'pihole-FTL': No such file
addr2line: 'pihole-FTL': No such file
addr2line: 'pihole-FTL': No such file
addr2line: 'pihole-FTL': No such file
addr2line: 'pihole-FTL': No such file
addr2line: 'pihole-FTL': No such file
```

**Comment by DL6ER** [comment]
> `sh: 1: addr2line: not found`

We call `addr2line` from within FTL when trying to generate a human-readable backtrace. It is a run-time dependency but only needed when FTL crashed. This dependency is there since at least one year ago, so we should definitely add it to the container. We just didn't realize, so far, because FTL never really crashed within a docker container (apparently).

> `addr2line: 'pihole-FTL': No such file`

This now means that `pihole-FTL` is not in `$PATH`. Here my knowledge about the inner workings of docker isn't sufficient.

---

From the last debug log, we see the crash happens in `dnsmasq` code somewhere in [this](https://github.com/pi-hole/FTL/blob/b90ab8b1e78d7aece11c8fee1c3c00817761de52/src/dnsmasq/util.c#L481-L505) subroutine. The issue is triggered by `listen-address=192.168.5.3`.

I have a proposed bugfix for this, but I'm not sure how to try custom FTL branches inside docker. Possibly by replacing the `pihole-FTL` binary temporarily? I'll defer this to @dschaper or @PromoFaux.

Branch `fix/bind-addresses`, PR https://github.com/pi-hole/FTL/pull/1132

**Comment by DL6ER** [comment]
Please also add the location of `pihole-FTL` to your `PATH` so `addr2line` can find it.

**Comment by DL6ER** [comment]
I do see extra log lines:
```
[2021-05-11 11:59:36.031 2370M] addr2line command: addr2line 0x5403e -e pihole-FTL
addr2line: 'pihole-FTL': No such file
```
Not sure what you were looking for.

Could you run those and give us the output?
```
addr2line 0x76144 -e pihole-FTL
addr2line 0x784bd -e pihole-FTL
addr2line 0x3f66c -e pihole-FTL
addr2line 0x3fc4a -e pihole-FTL
```

**Comment by lukaszczerpak** [comment]
ehhh apologies - indeed I wasn't sure what `addr2line` does.. now it's clear - lesson learnt :)

here you go:

```
# addr2line 0x76144 -e "$(which pihole-FTL)"
/usr/local/src/FTL/src/dnsmasq/dnsmasq.c:1820
# addr2line 0x784bd -e "$(which pihole-FTL)"
/usr/local/src/FTL/src/dnsmasq/dnsmasq.c:1227
# addr2line 0x3f66c -e "$(which pihole-FTL)"
/usr/local/src/FTL/src/main.c:98
# addr2line 0x3fc4a -e "$(which pihole-FTL)"
??:?
```

**Comment by DL6ER** [comment]
Sorry, I forgot the first (and, hence, most important) one...
```
addr2line 0x5403e -e "$(which pihole-FTL)"
```

**Comment by DL6ER** [comment]
I pushed another commit to the branch.

**Comment by PromoFaux** [comment]
> I have a proposed bugfix for this, but I'm not sure how to try custom FTL branches inside docker. Possibly by replacing the `pihole-FTL` binary temporarily?

Sorry, late to the party  here, but for future reference you should be able to:

 - make local clone of github.com/pi-hole/docker-pihole
 - checkout `dev` branch
 - change `FTL_VERSION=` in the file `VERSIONS` to `FTL_VERSION=fix/bind-addresses` (or whatever the branch name is)
 - run `gh-actions-test.sh` and let it build (obviously depends on docker being installed locally!)
 - Use the container that was just built (in my case on WSL it spits out `pihole/pihole:dev-amd64-buster`)
![image](https://user-images.githubusercontent.com/1998970/118016536-6dbeb300-b34d-11eb-96e8-caa5ece8a595.png)

**Comment by lukaszczerpak** [comment]
hi @DL6ER,

I can't see this PR https://github.com/pi-hole/FTL/pull/1132 merged into master. It's been done to development but never to master and thus new docker image doesn't contain it.
How to fix this?

**Comment by DL6ER** [comment]
> but never to master and thus new docker image doesn't contain it.

Yes, that's correct. There has been no release since we merged this. CoVid-19 and then the recent flooding in Germany affected us and things got delayed quite a bit.

>  How to fix this?

We are currently [beta testing](https://pi-hole.net/2021/07/04/pi-hole-ftl-v5-9-web-v5-6-and-core-v5-4-beta-testing/) the next version of Pi-hole. You can either join the beta testing (there is also a docker image for it) or wait for the next release. As we are synchronizing with `dnsmasq` about releases, we are not the ones to say: NOW so the best I can say is that should happen in the not too distant future (days to weeks).

**Comment by shaderecker** [comment]
This is fixed in the new `2021.09` release, right?
I can at least start up a container with a non-local ServerIP now.

**Comment by PromoFaux** [comment]
> This is fixed in the new `2021.09` release, right?

I think so! Will close out for now, we can always reopen.
